tests/test_nhg.py

```diff
@@ -384,20 +799,891 @@ def asic_route_nhg_fvs(k):
             route_ipprefix = gen_ipprefix(route_count)
             ps._del(route_ipprefix)
             route_count += 1
+        asic_routes_count -= MAX_ECMP_COUNT
 
         # Wait and check the second batch points to next hop group
         # Check ASIC DB on the count of nexthop groups used, and it should not increase or decrease
-        asic_db.wait_for_n_keys("ASIC_STATE:SAI_OBJECT_TYPE_NEXT_HOP_GROUP", 10)
-        keys = asic_db.get_keys("ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY")
-        k = asic_route_exists(keys, base_ipprefix)
+        asic_db.wait_for_n_keys(self.ASIC_NHG_STR, 10)
+        k = self.get_route_id(base_ipprefix, dvs)
         assert k is not None
         fvs = asic_route_nhg_fvs(k)
         assert fvs is not None
-        k = asic_route_exists(keys, last_ipprefix)
+        k = self.get_route_id(last_ipprefix, dvs)
         assert k is not None
         fvs = asic_route_nhg_fvs(k)
         assert fvs is not None
 
+        # Cleanup
+
+        # Remove second batch of routes
+        for i in range(10):
+            route_ipprefix = gen_ipprefix(MAX_ECMP_COUNT + i)
+            ps._del(route_ipprefix)
+
+        asic_db.wait_for_n_keys(self.ASIC_NHG_STR, 0)
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count)
+
+    def test_nhgorch_nhg_exhaust(self, dvs, testlog):
+        app_db = dvs.get_app_db()
+        asic_db = dvs.get_asic_db()
+        nhg_ps = swsscommon.ProducerStateTable(app_db.db_connection, "NEXTHOP_GROUP_TABLE")
+        rt_ps = swsscommon.ProducerStateTable(app_db.db_connection, "ROUTE_TABLE")
+
+        MAX_ECMP_COUNT = 512
+        MAX_PORT_COUNT = 10
+
+        r = 0
+        fmt = '{{0:0{}b}}'.format(MAX_PORT_COUNT)
+        asic_nhgs_count = len(asic_db.get_keys(self.ASIC_NHG_STR))
+        nhg_count = asic_nhgs_count
+        first_valid_nhg = nhg_count
+
+        def gen_nhg_fvs(binary):
+            nexthop = []
+            ifname = []
+
+            for i in range(MAX_PORT_COUNT):
+                if binary[i] == '1':
+                    nexthop.append(self.peer_ip(i))
+                    ifname.append(self.port_name(i))
+
+            nexthop = ','.join(nexthop)
+            ifname = ','.join(ifname)
+            fvs = swsscommon.FieldValuePairs([("nexthop", nexthop), ("ifname", ifname)])
+
+            return fvs
+
+        def gen_nhg_index(nhg_number):
+            return "group{}".format(nhg_number)
+
+        def gen_valid_binary():
+            nonlocal r
+
+            while True:
+                r += 1
+                binary = fmt.format(r)
+                # We need at least 2 ports for a nexthop group
+                if binary.count('1') <= 1:
+                    continue
+                return binary
```

Comment:
why not return a generator? it's cleaner.

Reply:
Fixed the nonlocal